Add support for VFS addons

notspiff opened the pull request:
This is my VFS add-on work. I send the PR to see if there is interest.

Some remaining stuff;
- get the datetime things into kodiplatform (see commit disabling it for now).
- clean up the callbacks. note; we have to do it like this due to threading stuff - we cannot just use the normal add-on callbacks.

see the repo list at https://github.com/notspiff for some example add-ons. No guarantee many of them are in working condition, i know vfs.sacd is (in terms of the iface), as well as vfs.rar, but there's some stability issues in the latter. also quite a few are now probably void since you have removed the code from mainline.

Jalle19 commented on xbmc/addons/Addon.cpp:
```diff
@@ -412,6 +413,8 @@ void CAddon::BuildLibName(const cp_extension_t *extension)
       case ADDON_SERVICE:
       case ADDON_REPOSITORY:
       case ADDON_CONTEXT_ITEM:
+      case ADDON_AUDIODECODER:
```
Where did this come from?

notspiff replied:
3 years of constant rebasing, where things originating in my branch has been pushed to mainline in slightly altered forms...

FernetMenta commented:
IMO an excellent idea. Once this is integrated we can start cleaning up vfs, i.e. PVR is the only client of LiveTV interface and can make use of this extension point.

FernetMenta commented on xbmc/cores/VideoPlayer/DVDInputStreams/DVDFactoryInputStream.cpp:
```diff
@@ -36,14 +36,25 @@
 #include "filesystem/File.h"
+#include "utils/StringUtils.h"
 #include "utils/URIUtils.h"
+#include "addons/VFSEntry.h"
```
Those changes here look like they don't belong to this PR.

notspiff replied:
they do, sort of. i have an add-on adding support for additional .iso files (SACD's). to distinguish these from the video iso images, the content type needs to be set on the item probed. if not it falls back to generic decision based on the fact that the extension is '.iso'.

FernetMenta replied:
why do you include AddonManager.h and VFSEntry.h? namespace ADDON is not used either.

notspiff replied:
ah, right. yes the includes should be removed.

Jalle19 commented on xbmc/addons/VFSEntry.cpp:
```diff
@@ -0,0 +1,666 @@
+  return false;
+}
+
+bool CVFSEntry::ContainsFiles(const CURL& url, CFileItemList& items)
```
It seems this method does a lot of different things, any chance it could be split up?

notspiff replied:
Not without changing the vfs api.

Jalle19 replied:
Okay

da-anda commented on addons/resource.language.en_gb/resources/strings.po:
```diff
@@ -876,7 +876,9 @@ msgctxt "#203"
 msgstr ""
 
-#empty string with id 204
+msgctxt "#204"
```
mind adding a contextual note and also the place where it's used? Something like:
```
#. Name of an add-on type/category
#:  xbmc/addons/Addon.cpp
```
Also, we prefer higher string IDs these days (5 digits), because lower numbers will give way to much wrong hits when searching the code for certain IDs. Thanks.

garbear commented on xbmc/addons/AddonCallbacksAddon.cpp:
```diff
@@ -543,6 +543,7 @@ static void CFileItemListToVFSDirEntries(VFSDirEntry* entries,
     entries[i].label = strdup(items[i]->GetLabel().c_str());
     entries[i].path = strdup(items[i]->GetPath().c_str());
     entries[i].size = items[i]->m_dwSize;
+    //entries[i].mtime = items[i]->m_dateTime;
```
Rebase error? The time field was removed from [kodi_vfs_types.h](https://github.com/xbmc/xbmc/blob/master/xbmc/addons/include/kodi_vfs_types.h)

notspiff replied:
uhm rewriting history? stuff came from me remember. why do you think it was there in the first place?

upstream in all its viceness removed my datetime stuff from kodiplatform. that is the only rebase error involved. it was removed from kodi_vfs_types because it was unused.

garbear commented on xbmc/addons/AddonManager.cpp:
```diff
@@ -37,6 +37,10 @@
+#include "utils/StringUtils.h"
+#include "utils/JobManager.h"
+#include "threads/SingleLock.h"
+#include "VFSEntry.h"
```
Looks like another rebase error. includes not needed

garbear commented on xbmc/addons/VFSEntry.cpp:
```diff
@@ -0,0 +1,666 @@
+  if (!Initialized())
+    return false;
+
+  return false;//m_pStruct->UpdateItem(ctx);
```
Intended?

garbear commented on xbmc/addons/VFSEntry.cpp:
```diff
@@ -0,0 +1,666 @@
+{
+  CVariant l2=0, l3=0;
+  if (line2)
+    l2 = std::string(line2);
```
is line2 null-checked before this line? I traced the execution back to the DLL, so we should sanity-check its values somewhere (probably as soon as it leaves the DLL boundary)

garbear commented:
@notspiff this PR was looking pretty good, what's the status? I'm happy to help if you have any TODOs left

notspiff commented:
i haven't touched it. there were some comments, feel free to address those. i will try to find some time..

notspiff committed:
add support for vfs add-ons

notspiff committed:
hack: keep a init level in service manager

VFS can be hit very early (InitDirectoriesXXX()) and with
vfs add-ons (and audio decoders) providing VFS entries,
we get in a squeeze.

this is not meant for inclusion as-is just highlights the issue.

notspiff committed:
disable mtime code for now

upstream kodiplatform is not prepared yet

notspiff committed:
hack: add iso to filedirectory list

notspiff commented:
rebased and updated. very little tested, in particular i'm not sure how well BinaryAddonCache manages the lifetime of add-ons.

example addon: https://github.com/notspiff/vfs.sacd

this is for SuperAudioCD isos. that is the reason for the last commit.

the hack commit highlights a nasty issue with current code design, which is triggered by adding this code.

garbear commented:
thanks for the rebase. it'll probably be a few months before I pull this into my retroplayer branch, i'm focusing on getting input near flawless for the v17 release.

notspiff commented:
Rather it goes to master. Retro carries enough /random as-is.

19. aug. 2016 17:17 skrev "Garrett Brown" <notifications@github.com>:

> thanks for the rebase. it'll probably be a few months before I pull this
> into my retroplayer branch, i'm focusing on getting input near flawless for
> the v17 release.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/xbmc/xbmc/pull/8966#issuecomment-241047181>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AFktYAkOZ8tuETGA3rkSlEZzukAbUwQ0ks5qhckigaJpZM4HNt6Q>
> .
>

garbear commented:
Rather RetroPlayer goes to master ;)

akva2 commented:
modernized https://github.com/notspiff/vfs.rar

added a commit that removes the internal rar support for this reason. the add-on will not work if core carries UnrarXLib.

akva2 commented:
modernized https://github.com/notspiff/vfs.sftp

added a commit that removes the internal sftp support.

bam80 commented:
What is left to merge it?

akva2 commented:
(notspiff under his day-work intentity); from my POV:

0) not being in a feature freeze
1) merging the awesome work of @AlwinEsch which cleans up the whole binary add-on concept in kodi, see https://github.com/AlwinEsch/kodi/pull/28
2) rewriting the API to use the new constructs and adding error handling and such (largely "automated" by using the new API constructs).
3) choosing a single add-on to focus on, preferably something that isn't new. like vfs.sftp. then remove remove the internal and have things real-world tested for a while. if all goes well, move more shit to add-ons / start adding new stuff like vfs.sacd

itouch5000 commented:
Any news on this topic? Will this change be part of Kodi 18?

notspiff commented:
rebased.

wsnipex commented:
jenkins build this with addons please

notspiff commented:
ish, sorry for bad timing. i was asked to remove the removal commits.

Paxxi commented:
jenkins build this with addons please

Paxxi commented:
no relevant failures in jenkins.

ksooo commented on xbmc/ServiceManager.h:
```diff
@@ -72,6 +72,7 @@ class CServiceManager
   CPlatform& GetPlatform();
 
   PLAYLIST::CPlayListPlayer& GetPlaylistPlayer();
+  int init_level = 0;
```
Is there a reason for this being a public member?

Why is the member decl placed randomly in between member function's decl instead of having it where the other members are? 

And why has it such a "strange" name? Not 100% sure, but don't we usually use the same notation for members, would be <code>iInitLevel</code> in this case?

ksooo replied:
> this is not meant for inclusion as-is just highlights the issue.

 hmm.

notspiff replied:
as i tried to explain in the commit message and the original PR description, this was meant to highlight the issue (using reference returns for things that is dynamically allocated). i just did what was necessary to make it work again and variable names was not a focus point ;)

notspiff replied:
oh, and the issue already exists in mainline, it just doesn't bite that often (you need to have a file type handled by a audio decoder with multitrack support sitting in some directory listed at early stages in the application). it bit me with rar cause i had some archive in a temp dir which made me notice..
